### PR TITLE
DO NOT MERGE: Pods with the DisruptionTarget condition should be treated as terminating by the endpoints controller

### LINKED ISSUE #116965: Graceful Node Shutdown does not update endpoints for terminating pods
### What happened?

When graceful node shutdown is enabled, I expect that services are drained appropriately during pod shutdown (i.e. the same way that `kubectl delete pod` would work). However this doesn't seem to be the case - the pod never enters the `Terminating` state according to `kubectl` and doesn't get removed from the service endpoints until it enters the `Completed` or `Failed` state. This causes connection errors and lost requests in `Cluster`-routed services.

### What did you expect to happen?

After pod termination starts, the EndpointSlice should be updated to have the pod that is terminating removed or marked terminating, even before the pod is fully shut down. This is what happens when a pod gets removed for any other reason (e.g. `kubectl delete pod`).

### How can we reproduce it (as minimally and precisely as possible)?

You can see the repro at https://github.com/mikekap/vagrant-kubeadm-kubernetes/tree/graceful-shutdown-repro . The repro sets up a 2 node cluster with graceful shutdown enabled and a simple deployment/service.

<details>
<summary>Here is the deployment/service yaml pasted again</summary>

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: echo-deployment
spec:
  selector:
    matchLabels:
      app: echo
  replicas: 2
  template:
    metadata:
      labels:
        app: echo
    spec:
      containers:
      - name: echo
        image: mendhak/http-https-echo
        readinessProbe:
          httpGet:
            path: /healthz
            port: 8080
        lifecycle:
          preStop:
            exec:
              command: ["/bin/sh", "-c", "echo 'Begin shutdown' > /proc/1/fd/1; sleep 20; echo 'End shutdown' > /proc/1/fd/1; exit 0"]
        ports:
        - containerPort: 8080
---
apiVersion: v1
kind: Service
metadata:
  name: my-service
spec:
  selector:
    app: echo
  ports:
    - protocol: TCP
      port: 80
      targetPort: 8080
```
</details>

To reproduce:
 * Run the deployment, preferably one pod on each node.
 * Run `watch -n 1 'kubectl  describe endpointslice my-service-5j2nw'` in a separate terminal (you'll need to figure out the right name unfortunately)
 * Use the virtualbox UI to send an ACPI shutdown to one of the nodes.

Within 20 seconds (the preStop hook duration), the EndpointSlice should be updated to have the pod that is terminating removed. This is what happens when a pod gets removed for any other reason (e.g. `kubectl delete pod`).

To ensure that everything else worked, you can reboot the node and see the journal logs (via something like `journalctl -b -1`) of the shutdown sequence triggering the kubelet graceful shutdown logic.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.2", GitCommit:"5835544ca568b757a8ecae5c153f317e5736700e", GitTreeState:"clean", BuildDate:"2022-09-21T14:33:49Z", GoVersion:"go1.19.1", Compiler:"gc", Platform:"darwin/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.3", GitCommit:"9e644106593f3f4aa98f8a84b23db5fa378900bd", GitTreeState:"clean", BuildDate:"2023-03-15T13:33:12Z", GoVersion:"go1.19.7", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
None
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.2 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.2 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
$ uname -a
Linux master-node 5.15.0-67-generic #74-Ubuntu SMP Wed Feb 22 14:14:39 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>
kubeadm
</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by smarterclayton**
Test a possible fix to the issue described in https://github.com/kubernetes/kubernetes/issues/116965#issuecomment-2195721381.

After discussion with @aojea and @mimowo and @alculquicondor, this has been updated to leverage (and slightly expand) the scope of the DisruptionTarget condition already used by JobController.

When the Kubelet expects to terminate a pod abnormally (i.e. not as a result of the pod exiting on its own or a deletionTimestamp), we use that info to always set a DisruptionTarget condition (and to override DisruptionTargets set by others).  We also correct a bug in Kubelet state transitions on termination that could delay sending condition changes unpredictably - the condition is now returned immediately after termination begins.  Since we now have a signal that the Kubelet intends to terminate a pod unexpectedly, the endpoints controller can safely use that condition to remove that pod from rotation, thus allowing network proxies the maximum amount of time to react to upcoming disruption.

TODO:
- [ ] Prove that this has no compatibility impact to existing workloads
- [ ] Identify a principle to guide addition of extra conditions if DisruptionTarget alone cannot cover the scenario
- [ ] Follow up to ensure probers continue running until the pods containers are stopped as discussed in #125558
- [ ] Separate the other commits into their own PRs - they are to aid in debugging or noticed while fixing
- [ ] Add better e2e suites around pod termination and ensure DisruptionTarget is set eagerly (at termination time) as well as simulate a graceful node shutdown and restart to ensure the condition is cleared
- [x] Add a reverse feature gate (on by default) to allow disabling the endpoints behavior if we find it breaks unexpectedly
- [ ] Look at the DisruptionTarget GC mechanism and modify it to account for the Kubelet now taking stronger ownership over reason: TerminatedByKubelet (i.e. we should let the kubelet own that state).
- [ ] Open a KEP PR that clarifies the behavior and expectations (to both DisruptionTarget and GracefulShutdown)

/kind bug

Fixes #116965, #124648

```release-note
Set the DisruptionTarget condition as soon as pods begin terminating.  Update the endpoints controller and endpointslice controller to consider a pod with DisruptionTarget to be "terminating", consistent with deletionTimestamp being set.  Add a feature gate `PodDisruptionConditionSignalsTerminating` that defaults to on but can be disabled if the feature causes impacts. Ensure the Kubelet removes the condition after a graceful node shutdown and reboot is completed.

This gives load balancers and proxies exposing the pods the maximum amount of time to remove pods from rotation before the Kubelet terminates the pod.
```
```docs
```

### CODE REVIEW

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -2176,7 +2176,7 @@ func (kl *Kubelet) SyncTerminatedPod(ctx context.Context, pod *v1.Pod, podStatus
 
 	// generate the final status of the pod
 	// TODO: should we simply fold this into TerminatePod? that would give a single pod update
-	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus, true)
+	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus, true, true)
```
Comment (aojea) [inline pkg/kubelet/kubelet.go:2179]: ok, this is clear, this has already terminated

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -2116,7 +2116,7 @@ func (kl *Kubelet) SyncTerminatingPod(_ context.Context, pod *v1.Pod, podStatus
 	// The computation is done here to ensure the pod status used for it contains
 	// information about the container end states (including exit codes) - when
 	// SyncTerminatedPod is called the containers may already be removed.
-	apiPodStatus = kl.generateAPIPodStatus(pod, podStatus, true)
+	apiPodStatus = kl.generateAPIPodStatus(pod, podStatus, true, true)
```
Comment (aojea) [inline pkg/kubelet/kubelet.go:2119]: ok, this is also clear, nothing running here

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -2039,7 +2039,7 @@ func (kl *Kubelet) SyncTerminatingPod(_ context.Context, pod *v1.Pod, podStatus
 	klog.V(4).InfoS("SyncTerminatingPod enter", "pod", klog.KObj(pod), "podUID", pod.UID)
 	defer klog.V(4).InfoS("SyncTerminatingPod exit", "pod", klog.KObj(pod), "podUID", pod.UID)
 
-	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus, false)
+	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus, true, false)
```
Comment (aojea) [inline pkg/kubelet/kubelet.go:2042]: this seems to be the important one

Comment (aojea) [reply pkg/kubelet/kubelet.go:2042]: readiness probes are removed on line 2068, before the pod is killed
```
	if err := kl.killPod(ctx, pod, p, gracePeriod); err != nil {
		kl.recorder.Eventf(pod, v1.EventTypeWarning, events.FailedToKillPod, "error killing pod: %v", err)
		// there was an error killing the pod, so we return that error directly
		utilruntime.HandleError(err)
		return err
	}
```

**Comment by aojea** [comment]
@smarterclayton I thought we were going to also enhance the endpoint slice controller to set the terminating condition 

https://github.com/kubernetes/kubernetes/blob/d902351c991a68fa76de9935a485afeb1f780c11/staging/src/k8s.io/endpointslice/utils.go#L37-L58

This will allow, from the network perspective of the cluster, to do the right things, workloads we know are going to be terminated because of reasons, can indicate they should only be used if there are no other workload ready

**Comment by smarterclayton** [comment]
>  thought we were going to also enhance the endpoint slice controller to set the terminating condition based on the Pod condition

https://github.com/kubernetes/kubernetes/pull/125774/files#diff-fc23208966763e11ddaea01723e972e3862da2f5a25d2f4c98e00b199655f23aR44

**Pull request closed without merging**